### PR TITLE
Grand Unified Peer Discovery

### PR DESCRIPTION
**Pull request by kim**
... which is a bold way of saying "we use DNS for peer discovery".

This turned out to be a bit more pedestrian than anticipated, mainly due to the `network` package being in a transitional phase (but us currently being unable to go straight to version 3.x), and some manual DNS juggling due to mDNS + testing (without modifying `/etc/resolv.conf`).

**tl;dr**:

* There is a new type and CLI parameter, `Network` (`--network`), which determines the logical oscoin network of the node ("mainnet", "testnet", etc.). This is used to minimise accidential joining of incompatible networks, but of course doesn't prevent it. The gossip handshake could ensure this in the future.
* As before, seed nodes can be given statically, if and when their IP addresses, host names, _and_ ports are known. Host names are resolved via DNS A/AAAA records, where _all_ returned addresses are considered as distinct nodes.
* It is possible to advertise a seed node farm (running on non-standard ports) via DNS SRV records. The format is fixed:

    `_gossip._tcp.oscoin.<network>.<search domain>`
    `_api._tcp.oscoin.<network>.<search domain>` 

   where `<network>` is the `Network` above, and `<search domain>` is a domain name owned by the node operator (e.g. "_gossip._tcp.oscoin.testnet.oscoin.io")
* A similar scheme can be enabled for "zeroconf" local area discovery via multicast DNS (`--enable-mdns`).

**Security Considerations**

DNSSEC is not currently supported, so using domain names is inherently insecure. Since we intend to rely on some form of authorisation during gossip handshake, this is of lesser concern.

### CODE REVIEW

File: test/Oscoin/Test/P2P/Disco.hs [hunk 1]
```diff
@@ -0,0 +1,234 @@
+module Oscoin.Test.P2P.Disco
+    ( tests
+    -- shut up, weeder!
```
Comment (cloudhead) [inline test/Oscoin/Test/P2P/Disco.hs]: I wasn't aware of this pragma! :joy:

**Comment by cloudhead** [comment]
Amazing! I can't really properly review this at the moment, I'd recommend you pair-review with @adinapoli-mndc or @geigerzaehler perhaps :dancing_men:

**Comment by kim** [comment]
@cloudhead Well, if you have a computer, perhaps spinning up a few nodes and eyeballing the logs to see if they find each other would be helpful (you can't do the SRV records of course, but permutations of `--seed` and `--enable-mdns`)

File: test/Oscoin/Test/P2P/Disco.hs (hunk 1 above)
Comment (kim) [reply test/Oscoin/Test/P2P/Disco.hs]: {-# LANGUAGE ShutUpIKnowBetter #-}

File: src/Oscoin/P2P/Types.hs [hunk 2]
```diff
@@ -32,12 +50,50 @@ import qualified Network.Gossip.IO.Peer as Gossip (Peer)
...
+import qualified Data.Vector as V
 import           Formatting as F
 import qualified Generics.SOP as SOP
-import           Network.Socket (HostName, PortNumber, SockAddr)
+import qualified Network.DNS as DNS
+import           Network.Socket (PortNumber, SockAddr)
+import qualified Network.Socket as Network
+
+-- | A logical oscoin network.
+data Network =
```
Comment (kim) [inline src/Oscoin/P2P/Types.hs]: This type should probably live somewhere else, as it is conceivable that other things may depend on it's value (in combination with `Environment`, supposedly)

File: src/Oscoin/P2P/Types.hs [hunk 3]
```diff
@@ -78,47 +205,19 @@ deriving instance Show (f (NodeId c)) => Show (NodeAddr f c)
...
-    p         <- note "Invalid port number" (readMaybe more)
-    pure NodeAddr
-        { nodeId   = Nothing
-        , nodeHost = h
-        , nodePort = p
-        }
-  where
-    readIP :: String -> Either String (IP, String)
-    readIP ('[' : more) =
+readNodeAddr = \case
```
Comment (kim) [inline src/Oscoin/P2P/Types.hs]: Future self may also make the port number optional here. We would need a contant to determine the default value, though, not hide it in the option parser for `main`

**Comment by cloudhead** [comment]
I tried running this locally, and here's what I get - let me know if I used the wrong flags:

Node 1 (no errors, uses keys in ./tmpkeys):
```
; stack exec -- oscoin --network testnet --enable-mdns --gossip-port 6943 --api-port 9999 --no-empty-blocks --keys tmpkeys
info  | running in env=development loc="Oscoin.Node:88:9"
info  | genesis is block_hash=2Drjgb6jvNEZTHjzLhZL81VtmF3kpuBUF7BN3hngBtCic2Pqwy loc="Oscoin.Node:91:9"
info  | mDNS: received query sender=172.20.2.35:51695 ns="disco" loc="Oscoin.P2P.Disco.MDns:114:7"
debug | mDNS: sending query recipient=172.20.2.35:51695 ns="disco" loc="Oscoin.P2P.Disco.MDns:114:7"
info  | mDNS: received response from=172.20.2.35:5152 ns="disco" loc="Oscoin.P2P.Disco.MDns:372:32"
info  | bootstrapping peer_addr=127.0.0.1:6943 peer_nodeid=2DrjgbDgeNKgzCuJvqWwQsKDHAzQ2XfyFvJCeBfm8j64QfTz3X ns="gossip" loc="Oscoin.P2P:134:10"
info  | bootstrapped peer_addr=127.0.0.1:6943 peer_nodeid=2DrjgbDgeNKgzCuJvqWwQsKDHAzQ2XfyFvJCeBfm8j64QfTz3X ns="gossip" loc="Oscoin.P2P:134:10"
Spock is running on port 9999
info  | mDNS: received query sender=172.20.2.35:57406 ns="disco" loc="Oscoin.P2P.Disco.MDns:114:7"
debug | mDNS: sending query recipient=172.20.2.35:57406 ns="disco" loc="Oscoin.P2P.Disco.MDns:114:7"
info  | incoming connection addr=127.0.0.1:38868 ns="gossip" loc="Oscoin.P2P:134:10"
debug | handshake complete peer_addr=127.0.0.1:38868 peer_nodeid=2Drjgb6UhVD1pFMVCUCH5x7MWbJMd6CNeSvn7KxYicuyM7LSTk ns="p2p" loc="Oscoin.P2P:170:13"
debug | received join peer_addr=127.0.0.1:38868 peer_nodeid=2Drjgb6UhVD1pFMVCUCH5x7MWbJMd6CNeSvn7KxYicuyM7LSTk ns="gossip" loc="Oscoin.P2P:134:10"
info  | connecting addr=127.0.0.1:6942 nodeid=2Drjgb6UhVD1pFMVCUCH5x7MWbJMd6CNeSvn7KxYicuyM7LSTk ns="gossip" loc="Oscoin.P2P:134:10"
info  | connected peer_addr=127.0.0.1:38868 peer_nodeid=2Drjgb6UhVD1pFMVCUCH5x7MWbJMd6CNeSvn7KxYicuyM7LSTk ns="gossip" loc="Oscoin.P2P:134:10"
debug | peer promoted to active view peer_addr=127.0.0.1:38868 peer_nodeid=2Drjgb6UhVD1pFMVCUCH5x7MWbJMd6CNeSvn7KxYicuyM7LSTk ns="gossip" loc="Oscoin.P2P:134:10"
debug | sending forward join peer_addr=127.0.0.1:38868 peer_nodeid=2Drjgb6UhVD1pFMVCUCH5x7MWbJMd6CNeSvn7KxYicuyM7LSTk ns="gossip" loc="Oscoin.P2P:134:10"
```

Node 2 (uses keys in ~/.config/oscoin):
```
; stack exec -- oscoin --network testnet --enable-mdns --no-empty-blocks              # 130
info  | running in env=development loc="Oscoin.Node:88:9"
info  | genesis is block_hash=2Drjgb6jvNEZTHjzLhZL81VtmF3kpuBUF7BN3hngBtCic2Pqwy loc="Oscoin.Node:91:9"
info  | mDNS: received response from=172.20.2.35:5152 ns="disco" loc="Oscoin.P2P.Disco.MDns:372:32"
info  | mDNS: received query sender=172.20.2.35:57406 ns="disco" loc="Oscoin.P2P.Disco.MDns:114:7"
debug | mDNS: sending query recipient=172.20.2.35:57406 ns="disco" loc="Oscoin.P2P.Disco.MDns:114:7"
info  | mDNS: received response from=172.20.2.35:5152 ns="disco" loc="Oscoin.P2P.Disco.MDns:372:32"
info  | bootstrapping peer_addr=127.0.0.1:6942 peer_nodeid=2Drjgb6UhVD1pFMVCUCH5x7MWbJMd6CNeSvn7KxYicuyM7LSTk ns="gossip" loc="Oscoin.P2P:134:10"
info  | connecting addr=127.0.0.1:6943 nodeid= ns="gossip" loc="Oscoin.P2P:134:10"
Spock is running on port 8477
debug | handshake complete peer_addr=127.0.0.1:6943 peer_nodeid=2DrjgbDgeNKgzCuJvqWwQsKDHAzQ2XfyFvJCeBfm8j64QfTz3X ns="p2p" loc="Oscoin.P2P:170:13"
info  | connected peer_addr=127.0.0.1:6943 peer_nodeid=2DrjgbDgeNKgzCuJvqWwQsKDHAzQ2XfyFvJCeBfm8j64QfTz3X ns="gossip" loc="Oscoin.P2P:134:10"
debug | peer promoted to active view peer_addr=127.0.0.1:6943 peer_nodeid=2DrjgbDgeNKgzCuJvqWwQsKDHAzQ2XfyFvJCeBfm8j64QfTz3X ns="gossip" loc="Oscoin.P2P:134:10"
debug | sending join peer_addr=127.0.0.1:6943 peer_nodeid=2DrjgbDgeNKgzCuJvqWwQsKDHAzQ2XfyFvJCeBfm8j64QfTz3X ns="gossip" loc="Oscoin.P2P:134:10"
info  | bootstrapped peer_addr=127.0.0.1:6942 peer_nodeid=2Drjgb6UhVD1pFMVCUCH5x7MWbJMd6CNeSvn7KxYicuyM7LSTk ns="gossip" loc="Oscoin.P2P:134:10"
error | Exception was thrown exception="Network.Socket.bind: resource busy (Address already in use)" callstack="CallStack (from HasCallStack):\n  logException, called at src/Oscoin/Telemetry/Logging.hs:312:63 in oscoin-0.1.0.0-B7A7fbBcskMGKPTvtnNLiR:Oscoin.Telemetry.Logging\n  withExceptionLogged, called at exe/node/Main.hs:142:11 in main:Main" loc="Main:142:11"
```

**Comment by cloudhead** [comment]
I ran it locally, here's what I found:
1. With `--no-empty-blocks` and `--enable-mdns` everything is smooth - ie. the handshake takes place and the nodes are connected.
2. If one node has `--no-empty-blocks` and not the other, I get:
```
error | P2P error error_class="deserialise_failure" error_message="Codec.CBOR: deserialising failed at offset 2 : Number of fields mismatch: expected=1 got=3" ns="p2p" loc="Oscoin.P2P:203:13"
```
3. Without `--no-empty-blocks` on either I get something that seems unrelated to P2P, but interesting:
```
info  | connecting addr=127.0.0.1:6942 nodeid= ns="gossip" loc="Oscoin.P2P:134:10"
Spock is running on port 9999
debug | handshake complete peer_addr=127.0.0.1:6942 peer_nodeid=2Drjgb6UhVD1pFMVCUCH5x7MWbJMd6CNeSvn7KxYicuyM7LSTk ns="p2p" loc="Oscoin.P2P:170:13"
info  | connected peer_addr=127.0.0.1:6942 peer_nodeid=2Drjgb6UhVD1pFMVCUCH5x7MWbJMd6CNeSvn7KxYicuyM7LSTk ns="gossip" loc="Oscoin.P2P:134:10"
debug | peer promoted to active view peer_addr=127.0.0.1:6942 peer_nodeid=2Drjgb6UhVD1pFMVCUCH5x7MWbJMd6CNeSvn7KxYicuyM7LSTk ns="gossip" loc="Oscoin.P2P:134:10"
debug | sending join peer_addr=127.0.0.1:6942 peer_nodeid=2Drjgb6UhVD1pFMVCUCH5x7MWbJMd6CNeSvn7KxYicuyM7LSTk ns="gossip" loc="Oscoin.P2P:134:10"
info  | bootstrapped peer_addr=127.0.0.1:6943 peer_nodeid=2DrjgbDgeNKgzCuJvqWwQsKDHAzQ2XfyFvJCeBfm8j64QfTz3X ns="gossip" loc="Oscoin.P2P:134:10"
error | Exception was thrown exception="SQLite3 returned ErrorBusy while attempting to perform step: database is locked" callstack="CallStack (from HasCallStack):\n  logException, called at src/Oscoin/Telemetry/Logging.hs:312:63 in oscoin-0.1.0.0-B7A7fbBcskMGKPTvtnNLiR:Oscoin.Telemetry.Logging\n  withExceptionLogged, called at exe/node/Main.hs:142:11 in main:Main" loc="Main:142:11"
```

One thing I notice also is when a peer is no longer reachable, the other one may also fail when trying to send, with:
```
info  | connecting addr=127.0.0.1:6943 nodeid=2DrjgbDgeNKgzCuJvqWwQsKDHAzQ2XfyFvJCeBfm8j64QfTz3X ns="gossip" loc="Oscoin.P2P:134:10"
info  | connected peer_addr=127.0.0.1:39048 peer_nodeid=2DrjgbDgeNKgzCuJvqWwQsKDHAzQ2XfyFvJCeBfm8j64QfTz3X ns="gossip" loc="Oscoin.P2P:134:10"
debug | peer promoted to active view peer_addr=127.0.0.1:39048 peer_nodeid=2DrjgbDgeNKgzCuJvqWwQsKDHAzQ2XfyFvJCeBfm8j64QfTz3X ns="gossip" loc="Oscoin.P2P:134:10"
debug | sending forward join peer_addr=127.0.0.1:39048 peer_nodeid=2DrjgbDgeNKgzCuJvqWwQsKDHAzQ2XfyFvJCeBfm8j64QfTz3X ns="gossip" loc="Oscoin.P2P:134:10"
debug | sending gossip peer_addr=127.0.0.1:39048 peer_nodeid=2DrjgbDgeNKgzCuJvqWwQsKDHAzQ2XfyFvJCeBfm8j64QfTz3X ns="gossip" loc="Oscoin.P2P:134:10"
error | error sending peer_addr=127.0.0.1:39048 peer_nodeid=2DrjgbDgeNKgzCuJvqWwQsKDHAzQ2XfyFvJCeBfm8j64QfTz3X exception="Network.Socket.ByteString.sendMany: invalid argument (Bad file descriptor)" ns="gossip" loc="Oscoin.P2P:134:10"
debug | peer demoted from active view peer_addr=127.0.0.1:39048 peer_nodeid=2DrjgbDgeNKgzCuJvqWwQsKDHAzQ2XfyFvJCeBfm8j64QfTz3X ns="gossip" loc="Oscoin.P2P:134:10"
info  | disconnected peer_addr=127.0.0.1:39048 peer_nodeid=2DrjgbDgeNKgzCuJvqWwQsKDHAzQ2XfyFvJCeBfm8j64QfTz3X ns="gossip" loc="Oscoin.P2P:134:10"
error | Exception was thrown exception="Network.Socket.ByteString.sendMany: invalid argument (Bad file descriptor)" callstack="CallStack (from HasCallStack):\n  logException, called at src/Oscoin/Telemetry/Logging.hs:312:63 in oscoin-0.1.0.0-B7A7fbBcskMGKPTvtnNLiR:Oscoin.Telemetry.Logging\n  withExceptionLogged, called at exe/node/Main.hs:142:11 in main:Main" loc="Main:142:11"
```

**Comment by cloudhead** [comment]
Interestingly also, two nodes using the same keypair have no problem connecting :smile:

File: src/Oscoin/P2P/Disco/MDns.hsc [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+            (msg,addr) <- receive sock
+            void . forkIO $ do
+                responderTrace rsp $ ResponderRecv msg addr
+                case filter ((== DNS.SRV) . DNS.qtype) $ DNS.question msg of
+                    q:_ ->
+                        case Map.lookup (DNS.qname q) answers of
+                            Nothing -> responderTrace rsp $
+                                ResponderHaveNoAnswer q addr
+                            Just rr -> respond sock addr $
+                                mkResponse lolRRs
```
Comment (adinapoli-mndc) [inline src/Oscoin/P2P/Disco/MDns.hsc]: Nice easter egg, sir. :trollface:

File: src/Oscoin/P2P/Disco/MDns.hsc [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+        , resolverTrace          = tracer
+        }
+  where
+    getRandom :: IORef ChaChaDRG -> IO Word16
+    getRandom ref = atomicModifyIORef' ref $ \gen ->
+      let
+          (bs, gen') = randomBytesGenerate 2 gen
+          !seqno     = case map fromIntegral (BS.unpack bs) of
+              (u:l:_) -> u * 256 + l
+              _       -> panic "To few random bytes"
```
Comment (adinapoli-mndc) [inline src/Oscoin/P2P/Disco/MDns.hsc]: Minor typo: s/To/Too

File: src/Oscoin/P2P/Disco/MDns.hsc [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+       in
+          (gen', seqno)
+
+data LookupOptions = LookupOptions
+    { optExpectPeers :: Int
+    -- ^ Expect at most this many peers to respond.
+    , optTimeout     :: Duration
```
Comment (adinapoli-mndc) [inline src/Oscoin/P2P/Disco/MDns.hsc]: Sometimes I wish we had:

```haskell

data MeasuredIn =
    Nanoseconds
  | Microseconds
  | Milliseconds
  | Seconds

newtype Duration (m :: MeasuredIn) = Duration Int64

-- optTimeout :: Duration 'Nanoseconds
```

😁 😁

File: src/Oscoin/P2P/Types.hs (hunk 2 above)
Comment (adinapoli-mndc) [reply src/Oscoin/P2P/Types.hs]: Brainstorming names here -- something like `Oscoin.Node.Infrastructure` ?

File: src/Oscoin/P2P/Types.hs [hunk 7]
```diff
@@ -32,12 +50,50 @@ import qualified Network.Gossip.IO.Peer as Gossip (Peer)
...
+import qualified Network.DNS as DNS
+import           Network.Socket (PortNumber, SockAddr)
+import qualified Network.Socket as Network
+
+-- | A logical oscoin network.
+data Network =
+      Mainnet
+    | Testnet
+    | Devnet
+    | Somenet Text
```
Comment (adinapoli-mndc) [inline src/Oscoin/P2P/Types.hs]: Nice, this is what I need in my [address format](https://github.com/oscoin/oscoin/pull/431/files#diff-23cbcb4fbca17bf5670558e55d30688cR72) patch, hehe.

File: test/Oscoin/Test/P2P/Disco.hs [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+
+tests :: TestTree
+tests = testGroup "Disco"
+    [ testProperty "prop_multicast"   prop_multicast
+    , testProperty "prop_staticPeers" prop_staticPeers
+    , testProperty "prop_dnsSD"       prop_dnsSD
+    ]
+
+{-
+props :: IO Bool
```
Comment (adinapoli-mndc) [inline test/Oscoin/Test/P2P/Disco.hs]: Needed anymore?

File: test/Oscoin/Test/P2P/Disco.hs [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+import           Network.Socket.ByteString (recvFrom, sendAllTo)
+
+import           Hedgehog
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import           Test.Tasty (TestTree, testGroup)
+import           Test.Tasty.Hedgehog (testProperty)
+
+tests :: TestTree
+tests = testGroup "Disco"
```
Comment (adinapoli-mndc) [inline test/Oscoin/Test/P2P/Disco.hs]: I can see we have quite a fair amount of hand-rolled parsers which can be easily turned into roundtrip specs of the form `renderFoo . parseFoo == id`. What do you think?

**Comment by adinapoli-mndc** [comment]
> on either I get something that seems unrelated to P2P, but interesting

@cloudhead that's because you definitely need to pass a `--blockstore` option to the nodes, otherwise they will try to write into the same DB, creating conflicts 😁 .


@kim, I have ran a number of different permutations of the commands using this tool called [hivemind](https://github.com/DarthSim/hivemind) which is rather nice: it doesn't require `tmux` and it allows you to write a `Procfile` and orchestrates things correctly. Below follows the content of the `Profile` for each run (in case you want to replicate locally) together with the output from the logs I'm getting:

### Experiment 1

```
node1: cabal new-run oscoin -- --network devnet --blockstore=node1.db --gossip-port 3000 --enable-mdns
node2: cabal new-run oscoin -- --network devnet --blockstore=node2.db --gossip-port 3001 --enable-mdns
``` 

Result:

[log1.txt](https://github.com/oscoin/oscoin/files/3017325/log1_clean.txt)

### Experiment 2

```
node1: cabal new-run oscoin -- --network devnet --blockstore=node1.db --gossip-port 3000 --enable-mdns
node2: cabal new-run oscoin -- --network devnet --blockstore=node2.db --gossip-port 3001 --seed 127.0.0.1:3000
```

Result:

[log2.txt](https://github.com/oscoin/oscoin/files/3017345/log2_clean.txt)

### Experiment 3

```
node1: cabal new-run oscoin -- --network devnet --blockstore=node1.db --gossip-port 9000 --no-empty-blocks --enable-mdns
node2: cabal new-run oscoin -- --network devnet --blockstore=node2.db --gossip-port 9001 --no-empty-blocks --enable-mdns
```

Result:

[log3.txt](https://github.com/oscoin/oscoin/files/3017376/log3_clean.txt)

**Comment by kim** [comment]
@adinapoli-mndc I think you need to specify `--api-port` and `--ekg-port` in the `Procfile`, the nodes are trying to bind to the same ones.

**Comment by adinapoli-mndc** [comment]
> I think you need to specify --api-port and --ekg-port in the Procfile, the nodes are trying to bind to the same ones.

Dang, I fell into the same fallacy I was advising against 😁 

(note my myself: it might be nice to have random port discovery for these services, even just for development purposes to avoid shooting ourselves in the foot so easily).

**Comment by adinapoli-mndc** [comment]
I do confirm that with:

```
node1: cabal new-run oscoin -- --network devnet --blockstore=node1.db --gossip-port 9000 --no-empty-blocks --enable-mdns --api-port 8000 --ekg-port 7000
node2: cabal new-run oscoin -- --network devnet --blockstore=node2.db --gossip-port 9001 --no-empty-blocks --enable-mdns --api-port 8001 --ekg-port 7001
```

It runs smooth as butter 😁

**Comment by kim** [comment]
> error | P2P error error_class="deserialise_failure"

Oops.

> you definitely need to pass a --blockstore option

Aww, wasn't aware of this either. 

> One thing I notice also is when a peer is no longer reachable, the other one may also fail when trying to send

Ah true. Someone ought to handle those exceptions 🙈 🙉 🙊 

> Interestingly also, two nodes using the same keypair have no problem connecting 

Yeah right... since the seeds don't have pre-shared keys, and their addresses are different, we only know when performing the handshake if the other side is using our key. Note that we can't at the moment prevent duplicate keys for the whole network. Kademlia queries the DHT when a node joins for the node id it identifies itself with -- but that is also a rather weak measure.

Interesting. I'm not sure - what would it mean for the correctness of the system if there were two different nodes with the same keypair in the network?

File: exe/node/Main.hs [hunk 10]
```diff
@@ -31,16 +33,18 @@ import           Oscoin.Telemetry.Metrics
...
 
 import           Options.Applicative
 
 data Args = Args
-    { host           :: HostName
+    { host           :: IP
     , gossipPort     :: PortNumber
     , apiPort        :: PortNumber
-    , seeds          :: [P2P.SeedAddr Crypto]
+    , discoOpts      :: P2P.Disco.Options
```
[the remark this answers is not in the document]
Comment (kim) [reply exe/node/Main.hs]: And... that’s a bad thing? :innocent:

[the remark this answers is not in the document]
Comment (cloudhead) [reply exe/node/Main.hs]: I think this is a feature, not a bug :smile:

**Comment by cloudhead** [comment]
> you definitely need to pass a --blockstore option
 
:man_facepalming: 

> Interesting. I'm not sure - what would it mean for the correctness of the system if there were two different nodes with the same keypair in the network?

Yeah this is what I've been wondering. Doesn't seem dangerous at first glance, but is there such a case where a packet is routed to a specific NodeId? If so, which one will it pick? And if both respond, what happens? etc.

**Comment by kim** [comment]
> is there such a case where a packet is routed to a specific NodeId?

No, peers are identified by their NodeId and network address.

I’m not too worried about the gossip layer, but wouldn’t it be problematic for two independent nodes to mine blocks with the same keypair?

File: exe/node/Main.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (adinapoli-mndc) [reply exe/node/Main.hs]: _Alfredo puts obligatory pun of nodes singing "staying alive, staying alive" when replying to DNS heartbeat queries. 😂 _

File: src/Oscoin/P2P/Disco.hs [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+import           Options.Applicative.Help
+                 (paragraph, stringChunk, unChunk, vcatChunks, vsepChunks)
+
+{-# ANN module ("HLint: ignore Use >=>" :: String) #-}
+
+data Options = Options
+    { optNetwork    :: Network
+    , optSeeds      :: [SeedAddr Crypto]
+    , optSDDomains  :: [HostName]
+    , optEnableMDns :: Bool
```
[the remark this answers is not in the document]
Comment (kim) [reply src/Oscoin/P2P/Disco.hs]: Yes please, make us one!

File: src/Oscoin/P2P/Disco/MDns.hsc (hunk 6 above)
Comment (kim) [reply src/Oscoin/P2P/Disco/MDns.hsc]: Yes please, make us one!

File: test/Oscoin/Test/P2P/Disco.hs (hunk 9 above)
Comment (kim) [reply test/Oscoin/Test/P2P/Disco.hs]: 2d3333b05822d1815b06c15dd41196708cbad1e1

File: test/Oscoin/Test/P2P/Disco.hs (hunk 8 above)
Comment (kim) [reply test/Oscoin/Test/P2P/Disco.hs]: Yeah, I took the liberty to expose those from all the test modules below `Oscoin.Test.P2P`, and teach `weeder` to shut up about unused exports once and for all. The type is `IO Bool`, which makes it very convenient to run just this test group in ghci.

/cc @cloudhead @geigerzaehler

File: src/Oscoin/P2P/Disco.hs [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+                    , "--sd-domain=monadic.xyz"
+                    ]
+                ]
+                )
+           <> metavar "DOMAIN NAME"
+            )
+        )
+    <*> switch
+        ( long "enable-mdns"
+       <> help "Enable mDNS discovery"
```
[the remark this answers is not in the document]
Comment (kim) [reply src/Oscoin/P2P/Disco.hs]: I'm reluctant -- this is more of a Use If You Know What You're Doing option at the moment.

File: src/Oscoin/P2P/Types.hs (hunk 2 above)
Comment (kim) [reply src/Oscoin/P2P/Types.hs]: https://github.com/oscoin/oscoin/issues/427#issuecomment-478059758

**Comment by kim** [comment]
@cloudhead @adinapoli-mndc PTAAL. There are, unsurpringly, some issues to be sorted out regarding the networking in general, but I think this PR accomplishes it's purpose.

**Comment by adinapoli-mndc** [comment]
@kim LGTM!

Started a cluster of 3 nodes with this `Procfile`:

```
node1: cabal new-run oscoin -- --network localhost --ekg-port 8001 --api-port 8002 --enable-mdns --blockstore store1.db
node2: cabal new-run oscoin -- --network localhost --ekg-port 9001 --api-port 9002 --enable-mdns --blockstore store2.db
node3: cabal new-run oscoin -- --network localhost --ekg-port 10001 --api-port 10002 --enable-mdns --blockstore store3.db
```

Which worked like a charm (log attached):

[log_clean.txt](https://github.com/oscoin/oscoin/files/3028494/log_clean.txt)

The only thing which would be cool to see I suppose would be log lines from the nodes when they actually receive block XYZ. I mean, something like this:

```
node1 | info  | broadcast block block_hash=2DrjgbFPGhFPMpp9RTQa1DSgyKJbF1FpqdQa4x87Dp61sp6aHa ns="node" loc="Oscoin.Node:129:52"

...

node2 | info  | received block block_hash=2DrjgbFPGhFPMpp9RTQa1DSgyKJbF1FpqdQa4x87Dp61sp6aHa ns="node" loc="Oscoin.Node:129:52"
```

But we can surely add that at the telemetry level as part of another PR 😉

**Comment by kim** [comment]
> log lines from the nodes when they actually receive block XYZ

But that _should_ actually be emitted: https://github.com/oscoin/oscoin/blob/master/src/Oscoin/P2P.hs#L225 🤔

**Comment by adinapoli-mndc** [comment]
True 🤔 That should appear only when logging with severity `Debug`, and with `env = development` that's indeed the verbosity level 🤔

**Comment by cloudhead** [comment]
> I’m not too worried about the gossip layer, but wouldn’t it be problematic for two independent nodes to mine blocks with the same keypair?

Good point - it shouldn't be an issue right now, given that we don't sign blocks, but if we want to do anything PoA-like, this will indeed be an issue. You could also argue that even in our "closed-network PoW" with a static key-set, you won't want a bunch of machines mining under the same NodeId, for fairness.

**Pull request merged**